### ISSUE TITLE
`--fix` with ignoring rule

### ISSUE THREAD

**Issue by azu** [textlint/textlint#211]
We have supported ignoring rule as experimental #193 

Currently, ignoring don't work with fixer.

This issue describe this bug. It is a bug of textlint-self.
https://github.com/textlint/textlint-rule-ignore-comments/issues/1

## Problem

Current, implementation of fixer.

> Parse text -> A fix rule -> re-parse text -> B fix rule ....

This design prefer much fixed than performance.

- https://github.com/textlint/textlint/blob/60633fe4530ee21e1da2d37386a568c5331f372c/src/fixer/fixer-processor.js#L36

**fixer and ignoring** 

Expected:

```
Parse text -> A fix rule -> re-parse text -> B fix rule ....
                  ^ + C ignoring rule                  ^ + C ignoring rule
```

Actual : 

> Parse text -> A fix rule -> re-parse text -> B fix rule -> re-parse text -> C ignoring rule ...

Not work ignoring.

It should be fixed

----

Edit(2016-05-26):

## Results after discussion

- Rename "ignoring rule" to "filter rule"
- Add "filters" filed support to textlint core
	- Same convention with "rules" field
- Rename "[textlint-rule-report-node-types](https://github.com/textlint/textlint-rule-report-node-types)" to "textlint-filter-rule-node-types". Is it better conventional name?
- Filter rule use `shouldIgnore(range, { ruleId })` API
  - https://github.com/textlint/textlint/blob/master/docs/rule.md
- textlint core still has actual filtering logic. - That is [message-filter.js](https://github.com/textlint/textlint/blob/master/src/shared/message-filter.js "message-filter.js")
  - [textlint-filter-plugin](https://github.com/textlint/textlint/issues/211#issuecomment-220991535) pattern is difficult.
  - We like simple.
- The order of filters doesn't affect result.
- textlint not allow (normal) rule to use `shouldIgnore`.

e.g.) The user want to use `textlint-filter-rule-node-types` and set following to  `.textlintrc` 

```json
{
   "filters": {
      "node-types": {
      	"nodeTypes": ["BlockQuote"]
      }
   }
}
```

See https://github.com/textlint/textlint/issues/211#issuecomment-221601511

**Comment by azu** [comment]
How to know textlint that rule is ignoring rule?

Pattern A

ignoring rule should export `ignoring`

```js
module.exports = {
    ignoring: reporter
};
```

Pattern B

Add `ignore` to `textlintrc` 

```js
{
   "ignore" : {  ... ignoreing rules ... }
   "rules": {}
}
```

**Comment by orangain** [comment]
As I don't know well about implementation of textlint core, fixers or ignoring rules, the following is just my opinion from view of a textlint user, not a developer:

At first, calling "ignoring rule" as "rule" seems to be a little bit strange because characteristic of ignoring rules somewhat differs from the one of other normal rules. Therefore, it seems that ignoring should be implemented by the textlint core or a plugin. If so, configuration should be like Pattern B. In pattern B, field name `ignore` is not so good because content of `ignore` field seems to be something should be ignored like `.gitignore`.

If ignoring rule is implemented by a rule as ever, I feel Pattern A is preferable. This is because I (as a user) don't want to configure rules depending on whether a rule supports ignoring or not.

**Comment by azu** [comment]
@orangain Thanks for feedback.

> calling "ignoring rule" as "rule" seems to be a little bit strange

Yes. I agree with this.

> Therefore, it seems that ignoring should be implemented by the textlint core or a plugin.

Yes and No.

I have experienced natural language/markup language as [domain-specific language (DSL)](https://en.wikipedia.org/wiki/Domain-specific_language "domain-specific language (DSL)").

For example, I write a [blog](http://jser.info/) that aggregate links of other sites.
I write markdown, but that markdown is a DSL.
It means that I not want to lint the title(`##`) by textlint.

```markdown
----
<!-- title come from the original site. I don't touch -->
## 6to5 · Turn ES6+ code into readable vanilla ES5
[6to5.org/](https://6to5.org/ "6to5 · Turn ES6+ code into readable vanilla ES5")

包含的にES6のコードをES5に変換する6to5の公式サイトリニューアル。
JSXやFlowとか多くの部分に対応、polyfill出来るところは外部のpolyfillを読み込み対応、可読性が高いコードを吐くようになっている。

Markdownやテキストだと、ここはこういうコンテキストで解釈するという独自解釈が多く存在すると思うので、そういうのコアで保証するのは難しい。
それぞれプラグイン作るのは今の仕組みだとイケてない…

----
```

So, I have created [textlint-plugin-jser-markdown](https://github.com/jser/textlint-plugin-jser-markdown "textlint-plugin-jser-markdown") as markdown processor plugin and use it.
Additionaly, exist similar cases I think.

In other word, textlint core or a single plugin fill the all need(ignoring error).

-----

Aside:

textlint define the plugin that has `preProcessor` and `postProcessor` method.
In other word, the plugin has `parser` and *filter*/modifier.

We can implemtned ignoring rule as pluing's `postProcessor`.
But, textlint use a single plugin per a file.

- https://github.com/textlint/textlint/blob/60633fe4530ee21e1da2d37386a568c5331f372c/src/util/proccesor-helper.js#L13

Currently, textlint can't apply multiple `postProcessor`.
It is become design issue... ?

We need to reconsider ignoring rule and plugin. 

-----

> If ignoring rule is implemented by a rule as ever

Yes and No. I've described this at the above(DSL).

**Comment by orangain** [comment]
Well, I understand that a plugin can ignore errors by example of textlint-plugin-jser-markdown. However, the plugin remove a node you want to ignore, so it seems to be hard to ignore only some specific rule. I feel more generic way like magic comments is preferable.

By your explanation, I realized what I really wanted to say. I think ignoring should be implemented as *another kind of* plugin, e.g. filter plugin, rather than a parser plugin.

My definition here is:

* parser plugin: parse some type of file, e.g. Markdown, Text, HTML, Re:VIEW, etc.
* filter plugin: filter or modify errors reported by rules. e.g. Ignore some problems depending on node types or magic comments.

-------

As you say, textlint's plugin has two functions. But, I guess, most of plugins mainly do its job only in `preProcessor`. So, how about splitting the two functions into the two kinds of plugins?

* A parser plugin implements `preProcessor`'s feature.
* A filter plugin implements `postProcessor`'s feature.
* A user can use only one parser plugin per a file and multiple filter plugins.

**Comment by azu** [comment]
> I think ignoring should be implemented as another kind of plugin

Yes! I called this as *filter* plugin in my head too. 💭 

> most of plugins mainly do its job only in preProcessor

I have known that two functions actural used in [textlint-plugin-satsuki](https://github.com/7pairs/textlint-plugin-satsuki "textlint-plugin-satsuki") by @7pairs.

- https://github.com/7pairs/textlint-plugin-satsuki/blob/f58d9ca7e7eb1475fe9cd6fc66088ee7d60020a1/src/SatsukiProcessor.js#L92-L114

Actually, textlint-plugin-satsuki use `postProcessor` for ignoring message.

Other plugins use only `preProcessor` as you said.

> So, how about splitting the two functions into the two kinds of plugins?

Yes, It is resonable idea. 💡 

But, current `postProcessor` have not `AST` and traverse function.
Ignoring rule need to `AST` and traverse function for ignoring where is.

- https://github.com/textlint/textlint-rule-ignore-comments/blob/a423c73a47f15bef80bec14d74090a219a89f69f/src/textlint-rule-ignore-comments.js#L40-L72

If *filter* plugin will implement own traverser, I think that it is a little difficult.

- https://github.com/textlint/textlint/blob/60633fe4530ee21e1da2d37386a568c5331f372c/src/task/textlint-core-task.js#L146-L162

We need to think about Filter plugin interface...

```js
// garbadge idea interface
class FilterPlugin {
    constructor(config){ }
    postProcessor like method()
    rules: { /* filter rules */ }
    rulesConfig: { /* rules config */ )
}
```

**Comment by orangain** [comment]
The idea of filter plugin sounds nice!

> I have known that two functions actural used in textlint-plugin-satsuki by @7pairs.

textlint-plugin-satsuki's use case of `postProcessor` is interesting. Using `postProcessor`, we can implement a new plugin without changing AST.

> But, current postProcessor have not traverse function that traverse TxtAST.

Sorry, I don't have enough knowledge to judge whether a filter plugin should implement traverser by itself or not.

> We need to think about Filter plugin interface...

Does FilterPlugin need `rules` and `rulesConfig`? I would imagine that the current implementation of textlint-rule-ignore-comments will be moved to a filter plugin like textlint-plugin-ignore-comments. So, I think the filter plugin does not need any rules. Configuration of the filter plugin will be done by `.textlintrc`.

**Comment by azu** [comment]
OK.

In brief

- Case 1: If _filter plugin_ have `rules` and `rulesConfig`, the plugin is the same with current implementation like this https://github.com/textlint/textlint-rule-ignore-comments
        - it is [storange](https://github.com/textlint/textlint/issues/211#issuecomment-220819392)
- Case 2: If _filter plugin_ not have `rules` and `ruleConfig`, the plugin should have to implement `traverse(AST)` and own filter.

> Does FilterPlugin need rules and rulesConfig?

Case 2 no needed `rules` and `rulesConfig`

@orangain Is your images like this?

```js
export default class FilterPlugin {
    /**
     * @param config config is textlint config
     */
    constructor(config) {
        this.config = config;
    }

    /**
     * This plugin filter `messages` using `AST`.
     *
     * @param messages
     * @param AST
     * @returns {Array} filtered messages
     */
    filter(messages, AST) {
        // This plugin implement the filter
        // messages ..... => filtered messages
        return messages;
    }
}
```

**Comment by azu** [comment]
I've implemented proof of concept of filter plugin.

- https://github.com/azu/textlint-filter-plugin-comments
- Notice: textlint not support this filter plguins. contain many hacks.

```js
const traverse = require("txt-ast-traverse").traverse;
// FIXME: a plugin/rule should not depended on textlint module
// It cause problem between textlint ver a and textlint ver b
// e.g.) 5.0 and 6.0 are duplicated installed... sad
const TextLintNodeType = require("textlint").TextLintNodeType;
const EventEmitter = require("events");
import ruleCreator from "./filter/textlint-rule-ignore-comments";
import StatusManager from "./filter/StatusManager";
// this copied from textlint core
import ignoringFilter from "./ignoring-filter"
export default class FilterPlugin {
    /**
     * @param config config is textlint config
     */
    constructor(config) {
        this.config = config;
    }

    /**
     * This plugin filter `messages` using `AST`.
     *
     * @param messages
     * @param AST
     * @returns {Array} filtered messages
     */
    filter(messages, AST) {
        // TODO: this has two issue.
        // 1. depended on textlint module
        // 2. use `raw` property(not recommended)
        const content = AST.raw;
        const statusManager = new StatusManager(content.length);
        const rule = ruleCreator({Syntax: TextLintNodeType, content, statusManager});
        const eventEmitter = new EventEmitter();
        // listen type and handler
        // This implementation is same textlint core
        Object.keys(rule).forEach(nodeType => {
            eventEmitter.on(nodeType, rule[nodeType]);
        });
        // traverse AST and found ignoring point
       // Note: implement own traverser
        traverse(AST, {
            enter(node) {
                eventEmitter.emit(node.type, node);
            },
            leave(node) {
                eventEmitter.emit(node.type + ":exit", node);
            }
        });
        // get result filtering point
        const ignoringMessages = statusManager.getIgnoringMessages();
        // filter messages by ignoringMessages
        return ignoringFilter(messages, ignoringMessages);
    }
}
```

This separation is fair, but it is very difficult to implement.
textlint core should provide helper to implement filter plugin...

**Comment by orangain** [comment]
Thanks for the concrete examples.

> Case 1: If filter plugin have rules and rulesConfig, the plugin is the same with current implementation like this https://github.com/textlint/textlint-rule-ignore-comments - it is storange

If textlint-rule-ignore-comments is renamed to textlint-filter-plugin-comments, it is **not** strange. In this case, interface of rule and filter plugin are almost the same from viewpoint of a developer. But from viewpoint of user, rule and filter plugin should have different names. I think (as a user) *the plugin* sounds like something which extends textlint's core features, but *the rule* sounds like something which checks usage of natural language.

In other word, the former is strange but the latter is not strange in the following `.textlintrc`.

```
# Feel strange
{
   "rules": {
       "no-todo",
       "ignore-comments"
   }
}

# Feel natural
{
   "plugins": {
        "filter-comments"
   },
   "rules": {
       "no-todo"
   }
}
```
Even in this case, I think `rules` and `rulesConfig` are not needed, because a rule does not have these properties. In my image, the filter plugin exports a function like the followings. This interface is the same as rule, but `context` may be an instance of `FilteringContext` which differs from `RuleContext`.

```js
export default function (context, options = defaultOptions) {
    const {Syntax, shouldIgnore, getSource} = context;

    const enablingComment = options.enablingComment || defaultOptions.enablingComment;
    const disablingComment = options.disablingComment || defaultOptions.disablingComment;

    const content = getSource();
    const statusManager = new StatusManager(content.length);

    return {
        [Syntax.Html](node){
            const nodeValue = getSource(node);
            if (!isHTMLComment(nodeValue)) {
                return;
            }
            const commentValue = getValueFromHTMLComment(nodeValue);
            if (commentValue.indexOf(enablingComment) !== -1) {
                const configValue = commentValue.replace(enablingComment, "");
                statusManager.enableReporting(node, parseRuleIds(configValue));
            } else if (commentValue.indexOf(disablingComment) !== -1) {
                const configValue = commentValue.replace(disablingComment, "");
                statusManager.disableReporting(node, parseRuleIds(configValue));
            }
        },
        [Syntax.Comment](node){
            const commentValue = node.value || "";
            if (commentValue.indexOf(enablingComment) !== -1) {
                const configValue = commentValue.replace(enablingComment, "");
                statusManager.enableReporting(node, parseRuleIds(configValue));
            } else if (commentValue.indexOf(disablingComment) !== -1) {
                const configValue = commentValue.replace(disablingComment, "");
                statusManager.disableReporting(node, parseRuleIds(configValue));
            }
        },
        [`${Syntax.Document}:exit`](){
            const ignoringMessages = statusManager.getIgnoringMessages();
            ignoringMessages.forEach(message => {
                const range = [message.startIndex, message.endIndex];
                shouldIgnore(range, {
                    ruleId: message.ruleId || "*"
                });
            })
        }
    }
};
```

----

> Case 2: If filter plugin not have rules and ruleConfig, the plugin should have to implement traverse(AST) and own filter.

> @orangain Is your images like this?

Yes. As you say, we need to consider the arguments of `filter()`.

> This separation is fair, but it is difficult to implement.
> textlint core should provide helper to implement filter plugin...

I agree with you. Some helper function to traverse tree with a Rule-like object may be helpful. Though it is just idea, how about using `FilteringContext` object which is similar to `RuleContext` but has an additional property `AST` and a method `traverse`?

```js
    /**
     * This plugin filter `messages` using `context`.
     *
     * @param messages
     * @param {FilteringContext} context
     * @returns {Array} filtered messages
     */
    filter(messages, context) {
        const content = context.getSource(context.AST);
        const statusManager = new StatusManager(content.length);
        const rule = ruleCreator({Syntax: context.Syntax, content, statusManager});

        context.traverse(rule);

        // get result filtering point
        const ignoringMessages = statusManager.getIgnoringMessages();
        // filter messages by ignoringMessages
        return ignoringFilter(messages, ignoringMessages);
    }
```

----

Finally, which to choice case 1 or 2 depends on whether a filter plugin needs more extensibility than rules. Unfortunately I don't have enough knowledge or experience to judge them.

**Comment by azu** [comment]
Thanks for comment!

📝 I notice that current "plugins" filed not support Options.
[textlint-rule-ignore-node-types](https://github.com/textlint/textlint-rule-ignore-node-types "textlint-rule-ignore-node-types") need option.

Case X: Babel-like pattern
```json
{
   "plugins": [{
      "plugin-a": { /* option */ }
   }]
}
```

Case Y:  add "filters" filed 

```js
{
   "filters": {
      "filter-a": { /* option */ }
   }
   // same convention with "rules"
}
```

**Comment by orangain** [comment]
> I notice that current "plugins" filed not support Options.

Oops! I didn't notice that.

I think "Case Y: add "filters" filed" will be preferable. This style is easy to learn because it has same convention with "rules" as you say.

However, I’m concerned that the order of filters to apply may possibly  affect in some cases (Though I have no concrete examples). If the order of filters doesn't affect so much, Case Y will be better.

----

Note that there could be another style:

Case Z: object-style "plugins"

```js
{
   "plugins": {
      "review": true,
      "filter-a": { /* option */ }
   }
   // same convention with "rules"
}
```

I think options will be useful even in a parser plugin. My story is as followings:

[Re:VIEW](https://github.com/kmuto/review) has [an interesting feature](https://github.com/kmuto/review/blob/master/doc/format.md#other-syntax) that a user can define custom blocks or inline tags using a ruby script file named `review-ext.rb`. [textlint-plugin-review](https://github.com/orangain/textlint-plugin-review) can't determine whether custom blocks should be parsed or ignored. In such a case, options to the plugin like `{"foo": "block", "bar": "ignore"}` will help to determine.

This is just my idea and there is no plan to implement them right now. But options to a parser plugin may be useful.

**Comment by azu** [comment]
> I think "Case Y: add "filters" filed" will be preferable. This style is easy to learn because it has same convention with "rules" as you say.

Yay! The **filter rule** <del>filter plugin</del> will be set in **"filters"** field.

I try to implement following:

- Add "filters" filed support to textlint core
	- Same convention with "rules" field
- Rename "textlint-rule-report-node-types" to "textlint-filter-rule-node-types". Is is better conventional name?
  - [textlint/textlint-rule-report-node-types: textlint rule that report error by node's type for debugging.](https://github.com/textlint/textlint-rule-report-node-types)
  - [textlint/textlint-rule-ignore-node-types: [Experimental] textlint rule that ignore node's type if the type is reported.](https://github.com/textlint/textlint-rule-ignore-node-types)
- textlint core has still actual filtering logic. - That is [message-filter.js](https://github.com/textlint/textlint/blob/master/src/shared/message-filter.js "message-filter.js")
  - [textlint-filter](https://github.com/textlint/textlint/issues/211#issuecomment-220991535) pattern is difficult.
  - We like simple.
- The order of filters doesn't affect result.

e.g.) The user want to use `textlint-filter-rule-node-types` and set following to  `.textlintrc` 

```json
{
   "filters": {
      "node-types": {
      	"nodeTypes": ["BlockQuote"]
      }
   }
}
```

@orangain What do you think about this?

----

> options to a parser plugin may be useful.

Yes. It will be required.
It will be separeted issue.

**Comment by orangain** [comment]
That's all OK :+1: The filter-rule sounds good!

**Comment by azu** [comment]
@orangain Thank you for giving me your cooperation 👍

**Comment by azu** [comment]
I've created proposal issue #212

**Comment by azu** [comment]
This issue( `--fix` with ignoring rule is not working ) is resolved by #214 
We will continue update follow todo https://github.com/textlint/textlint/issues/212#issuecomment-222353435
